exporter/prometheusexporter/metric_rename.go

```diff
@@ -0,0 +1,80 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheusexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusexporter"
+
+import (
+	"strings"
+
+	"go.opentelemetry.io/collector/pdata/pmetric"
+)
+
+var UnitMap = map[string]string{
+	"s":   "seconds",
+	"Cel": "celsius",
+	"m":   "meters",
+	"By":  "bytes",
+	"V":   "volts",
+	"A":   "amperes",
+	"J":   "joules",
+	"W":   "watts",
+	"g":   "grams",
+	"Hz":  "hertz",
+	"$":   "dollars",
+	"1":   "ratio",
+}
+
+// Build a Prometheus-compliant metric name for the specified metric
+func buildPromCompliantName(metric pmetric.Metric) string {
+
+	name := sanitize(strings.TrimSpace(metric.Name()), true)
+	unit := strings.TrimSpace(metric.Unit())
+	isCounter := metric.DataType() == pmetric.MetricDataTypeSum && metric.Sum().IsMonotonic()
+
+	// Remove _total suffix for now
+	// The _total suffix will be appended later for Counters
+	if isCounter {
+		name = strings.TrimSuffix(name, "_total")
```

Comment:
True, but metrics don't necessarily come from **prometheusreceiver** and we still want to make sure `_total` is removed for a *Counter*, so we can process the unit and then re-add `_total` at the end.

Reply:
Are there other formats where we expect `_total` to be a common suffix?

Reply:
Nope, not that I know of. Also, we're trimming it if and only if the metric is a *Monotonic Sum* (i.e. a *Counter*), in which case we will be appending `_total` after the unit.

Reply:
I'm fine with trimming _total if there are cases we know of in which metrics commonly have that suffix.  But i'd prefer to omit it from the initial version if we don't know of any cases.

Reply:
It's important to trim `_total` because we're appending it again for *Monotonic Sums*. Otherwise, we may end up going from `network.received_bytes_total` (with unit `By`) to `network_received_bytes_total_bytes_total`.

The same logic applies for other suffixes that we append to the metric name, like the unit: if it's already present in the metric name, we don't append it.

`_total` is a bit special because metrics names are typically structured as `<namespace>.<object>.<metric>[_<unit>][_total]`, so we need to remove it to ensure we check the `[_<unit>]` suffix, and then re-add `_total` for *Monotonic Sums*.

Reply:
> metrics names are typically structured as <namespace>.<object>.<metric>[_<unit>][_total]

Is that typical outside of Prometheus?  Thats what i'm trying to figure out.  If it is a prometheus metric, it is typically structured as <namespace>_<name>_<unit>_<total>, but we expect the prometheus receiver to handle this case by trimming it down to <namespace>_<name> before it gets to the exporter.

Reply:
Really I'm not comfortable in making **prometheusexporter** rely strongly on the behavior or **prometheusreceiver**. If the behavior of the receiver changes in the future, it will cause issues in users' environments that will go unnoticed in our unit and integration tests.

Also, there is the case where the flow of data is slightly different. If the metrics take a little detour through Splunk/SignalFX, for example: Prometheus exporter --> SignalFX Agent --> **signalfxreceiver** --> **prometheusexporter**.